sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs

```diff
@@ -22,6 +24,947 @@ namespace Azure.Messaging.EventHubs.Tests
+        ///   constructor.
+        /// </summary>
+        ///
+        [Test]
+        public void NamespaceConstructorCopiesTheIdentifier()
+        {
+            var options = new EventProcessorOptions { Identifier = Guid.NewGuid().ToString() };
+            var eventProcessor = new BasicProcessorMock(111, "consumerGroup", "namespace", "hub", Mock.Of<TokenCredential>(), options);
+
+            Assert.That(eventProcessor.Identifier, Is.EqualTo(options.Identifier));
+        }
+
+        /// <summary>
+        ///   Verifies functionality of the <see cref="EventProcessor{TPartition}.StartProcessing" />
+        ///   method.
+        /// </summary>
+        ///
+        [Test]
+        [TestCase(true)]
+        [TestCase(false)]
+        public void StartProcessingRespectsACanceledToken(bool async)
+        {
+            using var cancellationSource = new CancellationTokenSource();
+            cancellationSource.Cancel();
+
+            var mockProcessor = new Mock<EventProcessor<EventProcessorPartition>>(4, "consumerGroup", "namespace", "eventHub", Mock.Of<TokenCredential>(), default(EventProcessorOptions)) { CallBase = true };
+
+            if (async)
+            {
+                Assert.That(async () => await mockProcessor.Object.StartProcessingAsync(cancellationSource.Token), Throws.InstanceOf<TaskCanceledException>(), "The asynchronous call should have been canceled.");
+            }
+            else
+            {
+                Assert.That(() => mockProcessor.Object.StartProcessing(cancellationSource.Token), Throws.InstanceOf<TaskCanceledException>(), "The synchronous call should have been canceled.");
+            }
+        }
+
+        /// <summary>
+        ///   Verifies functionality of the <see cref="EventProcessor{TPartition}.StartProcessing" />
+        ///   method.
+        /// </summary>
+        ///
+        [Test]
+        [TestCase(true)]
+        [TestCase(false)]
+        public async Task StartProcessingStartsTheProcessing(bool async)
+        {
+            using var cancellationSource = new CancellationTokenSource();
+            cancellationSource.CancelAfter(TimeSpan.FromSeconds(15));
+
+            var mockProcessor = new Mock<EventProcessor<EventProcessorPartition>>(4, "consumerGroup", "namespace", "eventHub", Mock.Of<TokenCredential>(), default(EventProcessorOptions)) { CallBase = true };
+
+            mockProcessor
+                .Setup(processor => processor.CreateConnection())
+                .Returns(Mock.Of<EventHubConnection>());
+
+            if (async)
+            {
+                await mockProcessor.Object.StartProcessingAsync(cancellationSource.Token);
+            }
+            else
+            {
+                mockProcessor.Object.StartProcessing(cancellationSource.Token);
+            }
+
+            Assert.That(cancellationSource.IsCancellationRequested, Is.False, "The cancellation token should not have been signaled.");
```

Comment:
Question (and thanks in advance for teaching me things!):  In what cases would `cancellationSource.IsCancellationRequested` return true, yet we'd get to this line at all?  I would have assumed that if the cancellationToken fired, we'd observe either a `TaskCanceledException` or a `OperationCanceledException` from the call to `StartProcessing`?

Reply:
In our case, it would always  be a `TaskCanceledException`, as the Event Hubs documentation for errors calls out `OperationCancelledException` as having special meaning in the context.

Under normal circumstances, if everything is working correctly, you are right.  We *should* see an exception if cancellation is triggered.   If we got something wrong or something hang long enough to trigger the signal after we stopped checking the token, this assert is intended that we catch it.

We've had some interesting finds using this pattern, mostly in the live tests, but I'm a fan or verifying the assumption here as well.

Reply:
Looking at this with fresh eyes, I left off a couple of bits of potentially interesting context:

- The AMQP library that things are built-on has no concept of cancellation; its all timeout-based.  So cancellation checks and propagation is the responsibility of our code, which is why I was alluding to the potential for mistakes.

- During close/dispose/teardown, we deliberately ignore cancellation signals in many cases to avoid leaving things in an inconsistent state.  Its not likely, but possible, that we'd see a token signaled with no corresponding exception in that case.

Reply:
Thanks for explain this to me Jesse!  It makes a bunch of sense, and I agree with your position and reasoning.